### PR TITLE
Cache `ir_utils::allTvs` as part of Fusion

### PR DESCRIPTION
**Pull request by jacobhinkle**
This caches `ir_utils::allTvs(fusion)` as `fusion->allTvs()`. The cache is automatically invalidated whenever the TV graph topology changes; this mechanism is the same one used to recompute `Expr` uses automatically.

### CODE REVIEW

**Comment by jacobhinkle** [comment]
!build

**Comment by csarofeen** [comment]
Topo sort test was unhappy because it uses an unregistered expression and this approach triggers a traversal that needs dispatch. trying again.

File: csrc/fusion.cpp [hunk 1]
```diff
@@ -116,6 +116,14 @@ IrCloner Fusion::copy(const Fusion* from, Fusion* to) {
 
   to->expected_dynamic_smem_bytes_ = from->expected_dynamic_smem_bytes_;
 
+  if (from->all_tvs_ptr_ != nullptr) {
```
Comment (naoyam) [inline csrc/fusion.cpp:119]: nit: avoid unnecessary copy of a stale vector
```suggestion
  if (from->all_tvs_and_uses_valid_) {
```

File: csrc/fusion.h [hunk 2]
```diff
@@ -451,13 +455,14 @@ class NVF_API Fusion : public IrContainer {
...
-  //! Val::uses()
-  void resetTvUses();
+  //! Val::uses(). Also populate the all_tvs_ entry with all the used TVs.
+  void resetAllTvsAndUses();
 
   //! Declare that TensorView uses need to be updated (but don't actually do
   //! the update).
   void invalidateTvUses() {
-    all_tv_uses_valid_ = false;
+    all_tvs_and_uses_valid_ = false;
```
Comment (naoyam) [inline csrc/fusion.h:464]: This is an issue with the existing code, but this doesn't seem to be consistently used. We should just either remove it or rename it to something `invalidateAllTvsAndUses` and use it consistently.

**Comment by naoyam** [comment]
!build

**Comment by csarofeen** [comment]
!build

**Review by naoyam — APPROVED** [review]
LGTM

**Comment by naoyam** [comment]
!build

**Comment by csarofeen** [comment]
@naoyam let's hold off on merging this. I replaced all the uses of allTvs in the codebase and am now getting a segfault. There must still be something wrong with it.

**Comment by csarofeen** [comment]
> Do we still need to keep ir_utils::allTvs?

I'll cleanup in a follow up.

**Comment by csarofeen** [comment]
!build

**Pull request merged**